### PR TITLE
Feature/open new exam

### PR DESCRIPTION
**Pull request by millerts**

### CODE REVIEW

File: service/src/main/java/tds/exam/services/ConfigService.java [hunk 1]
```diff
@@ -12,9 +16,38 @@
...
+     * @param auditObject type of system flag
+     * @return {@link tds.config.ClientSystemFlag} if found otherwise empty
+     */
+    Optional<ClientSystemFlag> findClientSystemFlag(final String clientName, final String auditObject);
+
+    /**
+     * Finds the {@link tds.config.Accommodation} for the assessment
+     *
+     * @param assessmentKey the assessment key
+     * @return {@link tds.config.Accommodation} otherwise empty
```
Comment (millerts) [inline service/src/main/java/tds/exam/services/ConfigService.java]: fixed

File: service/src/main/java/tds/exam/repositories/impl/ExamCommandRepositoryImpl.java [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+            .addValue("assessmentId", exam.getAssessmentId())
+            .addValue("attempts", exam.getAttempts())
+            .addValue("status", exam.getStatus().getStatus())
+            .addValue("subject", exam.getSubject())
+            .addValue("studentKey", exam.getStudentKey())
+            .addValue("studentName", exam.getStudentName())
+            .addValue("browserId", getBytesFromUUID(exam.getBrowserId()))
+            .addValue("dateChanged", exam.getDateChanged())
+            .addValue("assessmentWindowId", exam.getAssessmentWindowId())
+            .addValue("segmented", exam.isSegmented() ? 1 : 0)
```
Comment (millerts) [inline service/src/main/java/tds/exam/repositories/impl/ExamCommandRepositoryImpl.java]: An oversight on my part

File: service/src/main/java/tds/exam/services/impl/SessionServiceImpl.java [hunk 3]
```diff
@@ -87,4 +84,23 @@ public SessionServiceImpl(RestTemplate restTemplate, ExamServiceProperties examS
...
+        UriComponentsBuilder builder =
+            UriComponentsBuilder.fromHttpUrl(String.format("%s/%s/assessment/%s", examServiceProperties.getSessionUrl(), sessionId, assessmentKey));
+
+        Optional<SessionAssessment> maybeSessionAssessment = Optional.empty();
+
+        try {
+            final SessionAssessment sessionAssessment = restTemplate.getForObject(builder.toUriString(), SessionAssessment.class);
+            maybeSessionAssessment = Optional.of(sessionAssessment);
+        } catch (HttpClientErrorException hce) {
+            if(hce.getStatusCode() != HttpStatus.NOT_FOUND) {
```
Comment (millerts) [inline service/src/main/java/tds/exam/services/impl/SessionServiceImpl.java]: Fixed

**Comment by millerts** [comment]
This has been updated per the in person code review.  @jtreuting , ready for your full review

File: service/src/main/java/tds/exam/services/impl/ExamServiceImpl.java [hunk 4]
```diff
@@ -336,35 +449,77 @@ public ExamServiceImpl(ExamQueryRepository examQueryRepository,
...
         return Optional.empty();
     }
+
+    //Should have long term cache
+    private boolean allowsGuestStudent(String clientName, ExternalSessionConfiguration externalSessionConfiguration) {
+        if (externalSessionConfiguration.isInSimulationEnvironment()) {
+            return true;
+        }
+
+        Optional<ClientSystemFlag> maybeSystemFlag = configService.findClientSystemFlag(clientName, ALLOW_ANONYMOUS_STUDENT_FLAG_TYPE);
```
Comment (jtreuting) [inline service/src/main/java/tds/exam/services/impl/ExamServiceImpl.java]: I feel like naming this `maybeAllowGuestAccessFlag` (or something similar) makes it easier to understand what that represents.  It feels like the systemFlag is almost a boolean and naming it kinda similarly seems appropriate.

File: service/src/main/java/tds/exam/services/impl/ExamServiceImpl.java [hunk 5]
```diff
@@ -336,35 +449,77 @@ public ExamServiceImpl(ExamQueryRepository examQueryRepository,
...
+
+        return maybeSystemFlag.isPresent() && maybeSystemFlag.get().isEnabled();
+    }
+
+    private void initializeExamAccommodations(Exam exam) {
+        //Replaces StudentDLL _InitOpportunityAccommodations_SP
+        Accommodation[] assessmentAccommodations = configService.findAssessmentAccommodations(exam.getAssessmentKey());
+
+        List<Accommodation> accommodations = Arrays.stream(assessmentAccommodations).filter(accommodation ->
```
Comment (jtreuting) [inline service/src/main/java/tds/exam/services/impl/ExamServiceImpl.java]: I think a comment here describing why the assessment accommodations are being filtered would help

File: service/src/test/java/tds/exam/services/impl/ExamServiceImplTest.java [hunk 6]
```diff
@@ -146,127 +205,269 @@ public void shouldReturnErrorWhenPreviousSessionTypeDoesNotEqualCurrentSessionTy
...
-        UUID sessionId = UUID.randomUUID();
-        OpenExamRequest openExamRequest = new OpenExamRequest();
-        openExamRequest.setStudentId(1);
-        openExamRequest.setSessionId(sessionId);
-        openExamRequest.setAssessmentId("assessmentId");
-        openExamRequest.setClientName("SBAC-PT");
-        openExamRequest.setMaxAttempts(5);
-
-        Session currentSession = new Session.Builder()
+    public void shouldOpenNewExamAsGuestAsGuest() {
```
Comment (jtreuting) [inline service/src/test/java/tds/exam/services/impl/ExamServiceImplTest.java]: DoubleDouble AsGuestAsGuest :)

File: service/src/test/java/tds/exam/services/impl/ExamServiceImplTest.java [hunk 7]
```diff
@@ -670,50 +911,50 @@ public void shouldGetInitialAbilityFromScoresForDifferentAssessment() {
...
-        ExternalSessionConfiguration externalSessionConfiguration = new ExternalSessionConfigurationBuilder().build();
-        when(sessionService.findExternalSessionConfigurationByClientName("SBAC-PT")).thenReturn(Optional.of(externalSessionConfiguration));
-
-        Response<Exam> examResponse = examService.openExam(openExamRequest);
-
-        assertThat(examResponse.getErrors()).isNotPresent();
-        assertThat(examResponse.getData()).isPresent();
-        assertThat(examResponse.getData().get().getId()).isEqualTo(previousExam.getId());
-    }
+//    @Test
```
Comment (jtreuting) [inline service/src/test/java/tds/exam/services/impl/ExamServiceImplTest.java]: Any reason to keep this in here?

File: service/src/main/java/tds/exam/repositories/impl/ExamAccommodationCommandRepositoryImpl.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+                .addValue("segmentKey", examAccommodation.getSegmentKey())
+                .addValue("type", examAccommodation.getType())
+                .addValue("code", examAccommodation.getCode())
+                .addValue("description", examAccommodation.getDescription())
+                .addValue("deniedAt", mapJodaInstantToTimestamp(examAccommodation.getDeniedAt()));
+
+            parameterSources.add(parameters);
+        });
+
+        jdbcTemplate.batchUpdate(SQL, parameterSources.toArray(new SqlParameterSource[parameterSources.size()]));
```
Comment (millerts) [inline service/src/main/java/tds/exam/repositories/impl/ExamAccommodationCommandRepositoryImpl.java]: Looks like something we can easily add when we make the db connection string.

File: service/src/main/java/tds/exam/repositories/impl/ExamCommandRepositoryImpl.java [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+            ":environment, " +
+            ":dateJoined " +
+            ")";
+
+
+        int insertCount = jdbcTemplate.update(SQL, parameters);
+
+        if (insertCount != 1) {
+            //TODO - should decide how to handle this.  If Exception probably should have a named runtime exception for this case
+            throw new RuntimeException(String.format("Failed to insert exam %s", exam));
```
Comment (millerts) [inline service/src/main/java/tds/exam/repositories/impl/ExamCommandRepositoryImpl.java]: Added a common CreateRecordException that is now thrown here.

File: service/src/main/java/tds/exam/services/impl/ConfigServiceImpl.java [hunk 10]
```diff
@@ -33,18 +38,70 @@ public ConfigServiceImpl(RestTemplate restTemplate, ExamServiceProperties examSe
     public Optional<ClientTestProperty> findClientTestProperty(final String clientName, final String assessmentId) {
         UriComponentsBuilder builder =
             UriComponentsBuilder
-                .fromHttpUrl(String.format("%s/%s/%s", examServiceProperties.getConfigUrl(), clientName, assessmentId));
+                .fromHttpUrl(String.format("%s/client-test-properties/%s/%s", examServiceProperties.getConfigUrl(), clientName, assessmentId));
```
Comment (millerts) [inline service/src/main/java/tds/exam/services/impl/ConfigServiceImpl.java]: I guess there is never an instance where we'd ever return all `client-test-properties` regardless of client?  Or really any property.  If that holds true I would imagine we should follow that paradigm everywhere.

I don't feel strongly about it either way except that clientname should always be the first path variable in the path for almost all URLs.

File: service/src/main/java/tds/exam/services/impl/ExamServiceImpl.java (hunk 4 above)
Comment (millerts) [inline service/src/main/java/tds/exam/services/impl/ExamServiceImpl.java]: Fixed.

File: service/src/test/java/tds/exam/services/impl/ExamServiceImplTest.java (hunk 6 above)
Comment (millerts) [inline service/src/test/java/tds/exam/services/impl/ExamServiceImplTest.java]: Fixed

File: service/src/test/java/tds/exam/services/impl/ExamServiceImplTest.java (hunk 7 above)
Comment (millerts) [inline service/src/test/java/tds/exam/services/impl/ExamServiceImplTest.java]: Not at this time.  I will add more tests when the previous exam work is complete.

File: service/src/main/java/tds/exam/services/impl/ConfigServiceImpl.java (hunk 10 above)
Comment (millerts) [inline service/src/main/java/tds/exam/services/impl/ConfigServiceImpl.java]: Regardless of the decision based on this discussion the changes will be in multiple projects and does impact the intent of this particular pull request.  For that reason I will make this particular change as a separate feature pull request.

File: service/src/main/java/tds/exam/services/impl/ExamServiceImpl.java (hunk 5 above)
Comment (millerts) [inline service/src/main/java/tds/exam/services/impl/ExamServiceImpl.java]: Improved the comments in the entire method referencing legacy code

File: service/src/test/java/tds/exam/services/impl/ExamServiceImplTest.java (hunk 7 above)
Comment (millerts) [inline service/src/test/java/tds/exam/services/impl/ExamServiceImplTest.java]: removed

**Pull request merged**